Add `Add customer` view

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #5288
Closes: #5123 
<!-- Id number of the GitHub issue this PR addresses. -->

~Please review and merge https://github.com/woocommerce/woocommerce-android/pull/5468 before checking this PR. Thanks!~

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds a new screen for `Add customer` screen that allows to fill billing address information with country and state included.

It also adds correct input types for address, making it easier for customers to add data.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

#### TC1 Regression tests

1. As this PR extracted address editing form, please smoke test order shipping and billing edditing.

#### TC2 Order Creation

1. Click `+` FAB on `Orders` tab
2. Click on `+ Add customer`
3. Scroll down
4. Click on `Country`
5. **Assert you can see and filter list of countries**
6. Click on `Algeria`
7. **Assert that you navigated back and you're on bottom of the view and Algeria is selected as country**
8. Click on `State`
5. **Assert you can see and filter list of states**
10. Select any state
7. **Assert that you navigated back and you're on bottom of the view and state is selected**
8. Click on `Country`
9. Select `Albania`
10. **Assert `State` is now editable `EditText`**

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->


https://user-images.githubusercontent.com/5845095/146531018-7f58104c-806a-4b13-b296-d71c1688f7f3.mov

| Before | After |
| ---- | ---- | 
| <img width="721" alt="Screenshot 2021-12-17 at 11 39 28" src="https://user-images.githubusercontent.com/5845095/146532045-ad278a33-46c3-4f24-b078-f8dd104e4b5c.png"> | <img width="721" alt="Screenshot 2021-12-17 at 11 33 23" src="https://user-images.githubusercontent.com/5845095/146532081-c993d493-7c9d-49d9-8469-d9352dcff574.png"> | 
| <img width="721" alt="Screenshot 2021-12-17 at 11 39 32" src="https://user-images.githubusercontent.com/5845095/146532144-97537e8b-541d-4e75-b542-2da1a5e9bb29.png"> | <img width="721" alt="Screenshot 2021-12-17 at 11 33 09" src="https://user-images.githubusercontent.com/5845095/146532197-709aaa89-7b7c-4372-81c8-9bbcc505abf8.png"> |





- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Refactor AddressViewModel to make it aplicable for Order Creation

wzieba committed:
Merge branch 'issue/5466_adjust_address_vm' into issue/5288_customer_form_order_creation

wzieba committed:
Extract Address edit form

wzieba committed:
Add Order Creation fragment

wzieba committed:
Connect navigation of Order Creation with Customer Creation screens

wzieba committed:
Merge remote-tracking branch 'origin' into issue/5288_customer_form_order_creation

wzieba commented on WooCommerce/src/main/res/layout/layout_address_form.xml:
```diff
@@ -0,0 +1,178 @@
+            android:layout_marginTop="@dimen/major_100"
+            android:hint="@string/order_detail_edit_address_details_first_name"
+            android:inputType="textPersonName|textCapWords"
+            android:nextFocusForward="@id/last_name"
```
`android:nextFocusForward` doesn't work (and didn't work before) because `WCMaterialOutlinedEditTextView` is not really an `EditTextView` but `TextInputLayout` which doesn't seem to support this option out of the box.

I've added a draft issue to our board to fix this and marked this as nice to have.

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationCustomerAddFragment.kt:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    @Suppress("FunctionOnlyReturningConstant")
+    private fun hasChanges() = true
```
To be done in scope of #5286

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationCustomerAddFragment.kt:
```diff
@@ -0,0 +1,136 @@
+        }
+    }
+
+    private fun showCountrySearchScreen() {
```
This and `showStateSearchScreen` is copied from `BaseAddressEditingFragment` with some differences like action direction and request key. 

I didn't want to extract this method as those two mentioned differences feels to be sufficient differences to not propose a shared method but I don't have strong feelings about this.

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationCustomerAddFragment.kt:
```diff
@@ -0,0 +1,136 @@
+        super.onViewCreated(view, savedInstanceState)
+        setHasOptionsMenu(true)
+
+        val billingBinding = LayoutAddressFormBinding.inflate(layoutInflater).apply {
```
Just recalled - I think I should nullify it in `onDestroyView`

wzieba replied:
Fixed in d599fe48bae02b7a5c7d9e2f3eff508028aa7bab

wzieba committed:
Merge remote-tracking branch 'origin' into issue/5288_customer_form_order_creation

wzieba committed:
Nullify billing binding on fragment destroy view

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationCustomerAddFragment.kt:
```diff
@@ -0,0 +1,139 @@
+
+    private var _billingBinding: LayoutAddressFormBinding? = null
+    private val billingBinding
+        get() = _billingBinding!!
```
Those don't seem to be accessed anywhere outside the scope of `onViewCreated`, so we should probably convert them to being local variables, and we'll be able to remove the `onDestroyView` override.

wzieba replied:
True, for now they're not. Fixed in 8149ba1a55b065de23ce1fabf499b4d57504d151

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationCustomerAddFragment.kt:
```diff
@@ -0,0 +1,139 @@
+        const val SELECT_BILLING_STATE_REQUEST = "select_billing_state_request"
+    }
+
+    private val addressViewModel by hiltNavGraphViewModels<AddressViewModel>(R.id.nav_graph_order_creations)
```
Do we need this to be navgraph scoped? I think the data will be saved to the `sharedViewModel`, which should be the source of truth for us, so this instance doesn't need to be retained when we leave the screen.

For more information check my comment below regarding navigateUp 🙂

wzieba replied:
Good point, thanks! I agree, it should be Fragment scoped. Fixed in eb2417db6e6c5e96d7ac56d37a7f96eb08efb7dd

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationCustomerAddFragment.kt:
```diff
@@ -0,0 +1,139 @@
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        return when (item.itemId) {
+            R.id.menu_done -> {
+                findNavController().navigateUp()
+                true
+            }
```
I believe we should communicate the updated Addresses to the `sharedViewModel` here?

wzieba replied:
True - I plan to add this in scope of a different task, #5286

hichamboushaba reviewed (commented):
Thanks for this @wzieba, I left some comments for discussion.

And I found one issue, I don't know if you are planning on fixing it later, since we don't call `start` of the ViewModel, we don't fetch the list of states/countries if they are missing.

wzieba committed:
Set fragment scope for AddressViewModel

wzieba committed:
Move bindings to local variables

wzieba commented:
Thanks @hichamboushaba for the review! I've addressed comments.

> And I found one issue, I don't know if you are planning on fixing it later, since we don't call start of the ViewModel, we don't fetch the list of states/countries if they are missing.

Yes, good point. I plan to call `start` and initialise ViewModel state in scope of #5290 - in the near future.

hichamboushaba reviewed (approved):
Thanks @wzieba, nice work.

There are some conflicts to be fixed before merging it 🙂

wzieba committed:
Merge remote-tracking branch 'origin' into issue/5288_customer_form_order_creation

wzieba commented:
Thanks @hichamboushaba ! I've just resolved conflicts, enabling auto-merge 🙂
